src/main/scala/loamstream/loam/LoamGraph.scala

```diff
@@ -201,12 +203,20 @@ final case class LoamGraph(
 
     val toolOutputsNew = toolOutputs + (tool -> (outputsFor(tool) -- stores))
 
-    val storeProducersNew = storeProducers.filterNot {
-      case (store, producer) => stores(store) && producer == tool
+    //Remove only at most `stores.size` keys from `storeProducers`.  Previously, this code relied on calling
+    //`storeProducers.filterNot`, which traversed the entire storeProducers Map and ran in O(nStores) time.
+    //The current approach is faster since for non-trivial pipelines, the number of keys to remove from 
+    //storeProducers is usually *much* smaller than the total number of stores, and removing a key from a map is
+    //fast, either O(1) or O(log nStores).
+    val storeProducersKeysToRemove: Iterable[Store] = {
+      stores.map { s => s -> storeProducers.get(s) }.collect { case (s, Some(producer)) if producer == tool => s }
```

Comment:
I _think_ I grok what the improvement is here... it went from:

```scala
bigListOfThings.filterNot(stores.contains(_))
```

to:

```scala
stores.reduce(bigListOfThings)(a, b => a.remove(b))
```

Where - because of the `BiMap`, `bigListOfThings` is now constant-time lookup.

If that's not the case, then the comment feels a bit off and it took me a bit to get there. But the comment does help. Just let me know if my conclusion was wrong. ;-)

The comment also states things like "is usually *much* smaller"; does this approach break if the number is not much smaller? What would be the worst-case big-O if the graph was not typical? I assume no worse than it currently is, but with words like "usually" it would be nice to get a better overall picture.

Reply:
> ```scala
bigListOfThings.filterNot(stores.contains(_))
```
> to:
> ```scala
stores.reduce(bigListOfThings)(a, b => a.remove(b))
```
Kinda, more or less.  The first way traverses a potentially huge Map (it has one entry for every Store) that records what Tool produces which Stores, so it's size is technically unbounded, but in big real-world pipelines is maybe 10s or 100s of thousands.  This traversal builds up a new potentially-huge map that's only smaller by (in practice) ~10s of entries at the most.  This matters because this method gets called (effectively) once for every tool!